src/Gui/Action.cpp

```diff
@@ -1243,4 +1350,206 @@ void WindowAction::addTo ( QWidget * w )
     }
 }
 
+// --------------------------------------------------------------------
+
+struct CmdInfo {
+    Command *cmd = nullptr;
+    QString text;
+    QString tooltip;
+    QIcon icon;
+    bool iconChecked = false;
+};
+static std::vector<CmdInfo> _Commands;
+static int _CommandRevision;
+
+class CommandModel : public QAbstractItemModel
+{
+public:
+
+public:
+    CommandModel(QObject* parent)
+        : QAbstractItemModel(parent)
+    {
+        update();
+    }
+
+    void update()
+    {
+        auto &manager = Application::Instance->commandManager();
+        if (_CommandRevision == manager.getRevision())
+            return;
+        beginResetModel();
+        _CommandRevision = manager.getRevision();
+        _Commands.clear();
+        for (auto &v : manager.getCommands()) {
+            _Commands.emplace_back();
+            auto &info = _Commands.back();
+            info.cmd = v.second;
+        }
+        endResetModel();
+    }
+
+    virtual QModelIndex parent(const QModelIndex &) const
+    {
+        return QModelIndex();
+    }
+
+    virtual QVariant data(const QModelIndex & index, int role) const
+    {
+        if (index.row() < 0 || index.row() >= (int)_Commands.size())
+            return QVariant();
+
+        auto &info = _Commands[index.row()];
+
+        switch(role) {
+        case Qt::DisplayRole:
+        case Qt::EditRole:
+            if (info.text.isEmpty()) {
+#if QT_VERSION>=QT_VERSION_CHECK(5,2,0)
+                info.text = QString::fromLatin1("%2 (%1)").arg(
+                        QString::fromLatin1(info.cmd->getName()),
+                        qApp->translate(info.cmd->className(), info.cmd->getMenuText()));
+#else
+                info.text = qApp->translate(info.cmd->className(), info.cmd->getMenuText());
+#endif
+                info.text.replace(QLatin1Char('&'), QString());
+                if (info.text.isEmpty())
+                    info.text = QString::fromLatin1(info.cmd->getName());
+            }
+            return info.text;
+
+        case Qt::DecorationRole:
+            if (!info.iconChecked) {
+                info.iconChecked = true;
+                if(info.cmd->getPixmap())
+                    info.icon = BitmapFactory().iconFromTheme(info.cmd->getPixmap());
+            }
+            return info.icon;
+
+        case Qt::ToolTipRole:
+            if (info.tooltip.isEmpty()) {
+                info.tooltip = QString::fromLatin1("%1: %2").arg(
+                        QString::fromLatin1(info.cmd->getName()),
+                        qApp->translate(info.cmd->className(), info.cmd->getMenuText()));
+                QString tooltip = qApp->translate(info.cmd->className(), info.cmd->getToolTipText());
+                if (tooltip.size())
+                    info.tooltip += QString::fromLatin1("\n\n") + tooltip;
+            }
+            return info.tooltip;
+
+        case Qt::UserRole:

```

Comment:
I think it's usually clearer to make a constant for the `Qt::UserRole` with a more descriptive name.